casanovo/utils.py

```diff
@@ -66,3 +74,191 @@ def split_version(version: str) -> Tuple[str, str, str]:
     """
     version_regex = re.compile(r"(\d+)\.(\d+)\.*(\d*)(?:.dev\d+.+)?")
     return tuple(g for g in version_regex.match(version).groups())
+
+
+def get_score_bins(
+    results_table: DataFrame, score_bins: List[float]
+) -> Dict[float, int]:
+    """
+    Get binned confidence scores
+
+    From a list of confidence scores, return a dictionary mapping each
+    confidence score to the number of spectra with a confidence greater
+    than or equal to it.
+
+    Parameters
+    ----------
+    results_table: DataFrame
+        Parsed spectrum match table
+    score_bins: List[float]
+        Confidence scores to map
+
+    Returns
+    -------
+    score_bin_dict: Dict[float, int]
+        Dictionary mapping each confidence score to the number of spectra
+        with a confidence greater than or equal to it.
+    """
+    return {
+        score: (results_table["score"] >= score).sum() for score in score_bins
+    }
+
+
+def get_peptide_length_histo(peptide_lengths: np.ndarray) -> Dict[int, int]:
+    """
+    Get a dictionary mapping each unique peptide length to its frequency
+
+    Parameters
+    ----------
+        peptide_lengths: np.ndarray
+            Numpy array containing the length of each sequence
+
+    Returns
+    -------
+        peptide_length_histogram: Dict[int, int]
+            Dictionary mapping each unique peptide length to its frequency
+    """
+    bins = np.bincount(peptide_lengths)
+    return {
+        curr_bin: count for curr_bin, count in enumerate(bins) if count != 0
+    }
+
+
+def get_peptide_lengths(results_table: DataFrame) -> np.ndarray:
+    """
+    Get a numpy array containing the length of each peptide sequence
+
+    Parameters
+    ----------
+    results_table: DataFrame
+        Parsed spectrum match table
+
+    Returns
+    -------
+    sequence_lengths: np.ndarray
+        Numpy array containing the length of each sequence, listed in the
+        same order that the sequences are provided in.
+    """
+    # Mass modifications do not contribute to sequence length
+    # If PTMs ae represented in ProForma notation this filtering operation
+    # needs to be reimplemented
+    alpha_re = re.compile("[^a-zA-Z]")
+    peptide_sequences = results_table["sequence"]
+    return peptide_sequences.str.replace(alpha_re, "", regex=True).apply(len)
+
+
+def get_report_dict(
+    results_table: DataFrame, score_bins: List[float] = SCORE_BINS
+) -> Optional[Dict]:
+    """
+    Generate sequencing run report
+
+    Parameters
+    ----------
+    results_table: DataFrame
+        Parsed spectrum match table
+    score_bins: List[float], Optional
+        Confidence scores for creating confidence CMF, see get_score_bins
+
+    Returns
+    -------
+    report_gen: Dict
+        Generated report represented as a dictionary, or None if no
+        sequencing predictions were logged
+    """
+    if results_table.empty:
+        return None
+
+    peptide_lengths = get_peptide_lengths(results_table)
```

Comment:
The functions themselves are short, but since the functionality of each of them is fairly unique imo combining them all into one function would make it difficult to thoroughly test each of them. Any thoughts on this would be appreciated.